### DISCUSSION TITLE
Use Cases

### DISCUSSION THREAD

**Discussion by Dolso** [AfterShip/clickhouse-sql-parser#35] [General]
Hello! I have a slightly abstract question.

If it's not a secret, can you please share what purpose this project is used for. For example, the item "Beautify ClickHouse SQL format" is completely understandable. And for example what problems do you solve with AST tree? Maybe some of the users told about variants of its interesting use?

Thanks in advance for the reply!

**Comment by git-hulk** [comment]
Hi @Dolso 

> And for example what problems do you solve with AST tree? Maybe some of the users told about variants of its interesting use?

We use the parsed AST to know what field and data internal developers use and access.

**Comment by hungtcs** [comment]
Hi @Dolso 

I have some uses that are different from Beautify, I am going to use it to modify the SQL entered by the user, as an example, let's say the query result now can't be more than 100 pieces of data at the most, I will convert the input SQL by modifying the AST like this

```sql
-- input
select * from users

-- output
with limit_wrapper as (
  select * from users
)
select * from limit_wrapper limit 100
```

Although the documentation does not explicitly state that it can be used for this purpose, it has worked well so far.

**Comment by vincentbernat** [comment]
Since it possible to transform an AST into a query, I think this could be a good low-level ORM. We could have a query builder building the AST, then get the query. This is more convenient/correct than using templating and this is more flexible than an ORM relying on models (like https://github.com/uptrace/go-clickhouse). Maybe the builder part could be part of this project?
